<a href="https://colab.research.google.com/github/scottwmwork/DS-Unit-1-Sprint-2-Data-Wrangling-and-Storytelling/blob/master/module1-join-and-reshape-data/LS_DS_121_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

_Lambda School Data Science_

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2019-08-13 00:01:30--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.133.45
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.133.45|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz.1’

instacart_online_gr 100%[===================>] 196.03M  34.4MB/s    in 6.0s    

2019-08-13 00:01:36 (32.7 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz.1’ saved [205548478/205548478]



In [0]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [0]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [0]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [0]:
##### YOUR CODE HERE #####
import pandas as pd

# What I need
#1.) Product name
#2.) Product ID
#3.) Order ID

In [0]:
#loading in dataframes
order_products__prior = pd.read_csv('order_products__prior.csv')
order_products__train = pd.read_csv('order_products__train.csv')
orders = pd.read_csv('orders.csv')
products = pd.read_csv('products.csv') 

In [0]:
order_products = pd.concat([order_products__prior, order_products__train])

In [0]:
#merge dataframe
newDataFrame = pd.merge(products,order_products, how = 'inner', on = 'product_id')

In [0]:
newDataFrame['product_name'].value_counts().head(10) #prints out the frequency of each variable 

Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
Name: product_name, dtype: int64

#From the data we now know: 

Banana  was ordered                  **491291** times

Bag of Organic Bananas  was ordered  **394930** times

Organic Strawberries was ordered     **275577** times

Organic Baby Spinach was ordered      **251705** times

Organic Hass Avocado was ordered      **220877** times 

Organic Avocado  was ordered          **184224** times 

Large Lemon was ordered               **160792** times

Strawberries was ordered              **149445** times

Limes  was ordered                    **146660** times

Organic Whole Milk  was ordered       **142813** times


## Reshape Data Section

- Replicate the lesson code
- Complete the code cells we skipped near the beginning of the notebook
- Table 2 --> Tidy
- Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [0]:
#oroginal code in this cell was broken, but I was able to find the dataset
flights = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/flights.csv')

In [0]:
##### YOUR CODE HERE #####
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

In [304]:
table1 = table1.reset_index()
table1

index  treatmenta  treatmentb
0    John Smith         NaN           2
1      Jane Doe        16.0          11
2  Mary Johnson         3.0           1

In [305]:
table1.columns.to_list()

['index', 'treatmenta', 'treatmentb']

In [306]:
tidy = table1.melt(id_vars='index')
tidy = table1.melt(id_vars='index', value_vars =['treatmenta','treatmentb'])
tidy

index    variable  value
0    John Smith  treatmenta    NaN
1      Jane Doe  treatmenta   16.0
2  Mary Johnson  treatmenta    3.0
3    John Smith  treatmentb    2.0
4      Jane Doe  treatmentb   11.0
5  Mary Johnson  treatmentb    1.0

In [307]:
tidy = tidy.rename(columns={
    'index': 'name', 
    'variable': 'trt', 
    'value': 'result'
})

tidy

name         trt  result
0    John Smith  treatmenta     NaN
1      Jane Doe  treatmenta    16.0
2  Mary Johnson  treatmenta     3.0
3    John Smith  treatmentb     2.0
4      Jane Doe  treatmentb    11.0
5  Mary Johnson  treatmentb     1.0

In [308]:
tidy.trt = tidy.trt.str.replace('treatment', '')

tidy

name trt  result
0    John Smith   a     NaN
1      Jane Doe   a    16.0
2  Mary Johnson   a     3.0
3    John Smith   b     2.0
4      Jane Doe   b    11.0
5  Mary Johnson   b     1.0

In [348]:
### Table 1 ---> Tidy
table_1 = tidy
table_1.trt = table_1.trt.str.replace('a', 'treatmenta')
table_1.trt = table_1.trt.str.replace('b', 'treatmentb')
table_1 = table_1.pivot(index = 'name', columns = 'trt', values = 'result')
table_1.rename(columns={
    'name': ' ', 
    'trt' : ' '
})

#Couldn't figure out how to remove trt and name

trt           tretreatmentatmenttreatmenta  tretreatmentatmenttreatmentb
name                                                                    
Jane Doe                              16.0                          11.0
John Smith                             NaN                           2.0
Mary Johnson                           3.0                           1.0

In [0]:
table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

In [354]:
table2 = table1.T
table2


John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [355]:
### Table 2 --> Tidy ###
table2 = table2.T #transpose it back
table2 = table2.reset_index() 
tidy = table2.melt(id_vars='index')
tidy = table2.melt(id_vars='index', value_vars =['treatmenta','treatmentb'])
 tidy

index    variable  value
0    John Smith  treatmenta    NaN
1      Jane Doe  treatmenta   16.0
2  Mary Johnson  treatmenta    3.0
3    John Smith  treatmentb    2.0
4      Jane Doe  treatmentb   11.0
5  Mary Johnson  treatmentb    1.0

In [357]:
# Load seaborn's flights dataset.
# Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. 
# You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.
flights.head(12)

index  year      month  passengers
0       0  1949    January         112
1       1  1949   February         118
2       2  1949      March         132
3       3  1949      April         129
4       4  1949        May         121
5       5  1949       June         135
6       6  1949       July         148
7       7  1949     August         148
8       8  1949  September         136
9       9  1949    October         119
10     10  1949   November         104
11     11  1949   December         118

In [366]:
flights_pivot = flights.pivot(index = 'month', columns = 'year', values = 'passengers')
flights_pivot #display new dataframe

year       1949  1950  1951  1952  1953  ...  1956  1957  1958  1959  1960
month                                    ...                              
April       129   135   163   181   235  ...   313   348   348   396   461
August      148   170   199   242   272  ...   405   467   505   559   606
December    118   140   166   194   201  ...   306   336   337   405   432
February    118   126   150   180   196  ...   277   301   318   342   391
January     112   115   145   171   196  ...   284   315   340   360   417
July        148   170   199   230   264  ...   413   465   491   548   622
June        135   149   178   218   243  ...   374   422   435   472   535
March       132   141   178   193   236  ...   317   356   362   406   419
May         121   125   172   183   229  ...   318   355   363   420   472
November    104   114   146   172   180  ...   271   305   310   362   390
October     119   133   162   191   211  ...   306   347   359   407   461
September   136   158   184   209   237  ...   355   404   404   463   508

[12 rows x 12 columns]

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####